systems/trajectory_optimization/dircon_options.cc

```diff
@@ -1,30 +1,111 @@
 #include "systems/trajectory_optimization/dircon_options.h"
 
+using std::unordered_map;
 using std::vector;
 
 namespace dairlib {
 namespace systems {
 namespace trajectory_optimization {
 
-DirconOptions::DirconOptions(int n_constraints) {
-  n_constraints_ = n_constraints;
-  is_constraints_relative_ = vector<bool>(n_constraints_);
-  for (int i=0; i < n_constraints_; i++) {
-    is_constraints_relative_[i] = false;
-  }
+DirconOptions::DirconOptions(int n_kin_constraints) {
+  n_kin_constraints_ = n_kin_constraints;
+  is_constraints_relative_ = vector<bool>(n_kin_constraints_, false);
   start_constraint_type_ = DirconKinConstraintType::kAll;
   end_constraint_type_ = DirconKinConstraintType::kAll;
   force_cost_ = 1.0e-4;
 }
+DirconOptions::DirconOptions(int n_constraints,
+                             drake::multibody::MultibodyPlant<double>* plant)
+    : DirconOptions(n_constraints) {
+  n_v_ = plant->num_velocities();
+  n_x_ = plant->num_positions() + plant->num_velocities();
+}
+DirconOptions::DirconOptions(
+    int n_constraints,
+    drake::multibody::MultibodyPlant<drake::AutoDiffXd>* plant)
+    : DirconOptions(n_constraints) {
+  n_v_ = plant->num_velocities();
+  n_x_ = plant->num_positions() + plant->num_velocities();
+}
+
+void DirconOptions::setDynConstraintScaling(double scale, int row_start,
+                                            int row_end) {
+  DRAKE_DEMAND(row_end < n_x_);
+  addConstraintScaling(&dyn_constraint_scaling_, scale, row_start, row_end);
+}
+void DirconOptions::setKinConstraintScaling(double scale, int row_start,
+                                            int row_end) {
+  DRAKE_DEMAND(row_end < 3 * n_kin_constraints_);
+  for (int i = row_start; i <= row_end; i++) {
+    if (i < n_kin_constraints_) {
```

Comment:
I'm still confused by this logic. Please explain in comments what you're doing and why. It seems like a pretty arcane `if else` construction that isn't at all obvious to a user. My suspicion is that this should be 3 functions, but without know what/why you're doing it this way, it's hard to tell.

Reply:
Based on my observation, the ratio of the scaling factors for each element in the position level to those in velocity/acceleration levels is roughly the same. For example, let's say there are two kinematics constraints. In DIRCON, we impose the constraint also in velocity and acceleration, so `n_kin` becomes 6. And my observation is that scaling_factor[0]/scaling_factor[2] is roughly the sample as scaling_factor[1]/scaling_factor[3]. 

This is why I scale everything in the velocity level by `kin_constraint_scaling_vel_`, and everything in position level by `kin_constraint_scaling_pos_`.

I think this is not easy to explain to the new users and might complicate the code in a unnecessary way, so I'm changing the API such that the users have to specify pos/vel/accel levels individually.